src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -406,57 +402,76 @@ void probe(const RadixContainer<RightType>& radix_container,
           }
 
-          const auto& rows_iter = hashtable.find(type_cast<HashedType>(row.value));
+          const auto& rows_iter = hash_table.find(type_cast<HashedType>(right_row.value));
 
-          if (rows_iter != hashtable.end()) {
-            // Key exists, thus we have at least one hit
-            const auto& matching_rows = rows_iter->second;
+          if (rows_iter != hash_table.end()) {
+            // Key exists, thus we have at least one hit for the first predicate
+            const auto& primary_predicate_matching_rows = rows_iter->second;
 
             // Since we cannot store NULL values directly in off-the-shelf containers,
             // we need to the check the NULL bit vector here because a NULL value (represented
             // as a zero) yields the same rows as an actual zero value.
             // For inner joins, we skip NULL values and output them for outer joins.
-            // Note, if the materialization/radix partitioning phase did not explicitely consider
+            // Note, if the materialization/radix partitioning phase did not explicitly consider
             // NULL values, they will not be handed to the probe function.
             if constexpr (consider_null_values) {
               if ((*radix_container.null_value_bitvector)[partition_offset]) {
                 if (mode == JoinMode::Left || mode == JoinMode::Right) {
                   pos_list_left_local.emplace_back(NULL_ROW_ID);
-                  pos_list_right_local.emplace_back(row.row_id);
+                  pos_list_right_local.emplace_back(right_row.row_id);
                 }
                 // ignore found matches and continue with next probe item
                 continue;
               }
             }
 
-            // If NULL values are discarded, the matching row pairs will be written to the result pos lists.
-            for (const auto& row_id : matching_rows) {
-              pos_list_left_local.emplace_back(row_id);
-              pos_list_right_local.emplace_back(row.row_id);
+            // If NULL values are discarded, the matching right_row pairs will be written to the result pos lists.
+            if (!multi_predicate_join_evaluator) {
+              for (const auto& row_id : primary_predicate_matching_rows) {
+                pos_list_left_local.emplace_back(row_id);
+                pos_list_right_local.emplace_back(right_row.row_id);
+              }
+            } else {
+              auto match_found = false;
+              for (const auto& row_id : primary_predicate_matching_rows) {
+                if (multi_predicate_join_evaluator->fulfills_all_predicates(row_id, right_row.row_id)) {
+                  pos_list_left_local.emplace_back(row_id);
+                  pos_list_right_local.emplace_back(right_row.row_id);
+                  match_found = true;
+                }
+              }
+              // We have not found matching items for all predicates.
+              if constexpr (consider_null_values) {
```

Comment:
"consider"? is that the same as "retain" as in "AntiRetainNulls"? then I'd call this variable "retain_null_values"